### ISSUE TITLE
core dumped occured with a leaf modification of a list in container

### ISSUE THREAD

**Issue by myonlystarWang** [sysrepo/sysrepo#2216]
Hi
     I modify a leaf value of a list in container from my webpage which can trigger netconf message to netopeer2-server,then trigger a sysrepo callback.But the core dumped happened by accident like following logs:
```
/home/user/new_work/sysrepo-1.4.58/src/shm_mod.c:430: sr_shmmod_modinfo_rdlock: Assertion `!shm_lock->write_locked' failed.
Aborted (core dumped)
```
Then I try to gdb the application but an error has been reported as follows :
```
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff6b8b700 (LWP 2784)]
[New Thread 0x7ffff638a700 (LWP 2785)]
[New Thread 0x7ffff5987700 (LWP 2786)]

[ERR]: Waiting on a conditional variable failed (sr_connect: Connection timed out).
[Thread 0x7ffff5987700 (LWP 2786) exited]
[Thread 0x7ffff638a700 (LWP 2785) exited]
[Thread 0x7ffff6b8b700 (LWP 2784) exited]
[Inferior 1 (process 2780) exited with code 01]
```

This does not happen every time I modify the list data,but it can only be recovered by reboot system

**Comment by michalvasko** [comment]
What *sysrepo* version are you using? I think it was already fixed.

**Comment by myonlystarWang** [comment]
Hi michal,
    Thanks for response.
    The using version is 1.4.58,but I have installed the latest devel verison in a new VM.
    I will have a full test with this operation,then see the result.
BR,
wang

**Comment by myonlystarWang** [comment]
@michalvasko 
Hi,
It occurred again with the devel version
```
root@user:/home/user# sysrepoctl -V
sysrepoctl - sysrepo YANG schema manipulation tool, compiled with libsysrepo v1.4.90 (SO v5.6.6)
```
and my application architecture is there threads,and one thread subscribe module change.I have installed a yang module which has one container as following:
```
module test{
    namespace "urn:hysatcom:params:xml:ns:yang:test";
	prefix test;
	
	revision 2020-09-01 {
		description
		    "initial create";
	}

	typedef my-decimal {
         type decimal64 {
             fraction-digits 16;
          }
       }
	 
	container config {
		config true;
					
		leaf param1 {
			type uint8 {
				range "1..15";
			}
		}
		
		leaf param2 {
			type my-decimal;
		}	

		list list1{
			key "list1Para1";
						
			leaf list1Para1 {
				type uint8 {
					range "1..3";
				}
			}
			
			leaf list1Para2{
				type my-decimal;
			}		
		}	
		list list2{
			key "list2Para1";
						
			leaf list2Para1{
				type uint8 {
					range "1..3";
				}
			}
			
			leaf list2Para2{
				type my-decimal;
			}		

			leaf list2Para3{
				type uint32;
			}	
		}	
	} //container config
```
my subscribe code is:
 ```
   
    iRet = sr_module_change_subscribe(session, "test", "/test:config", config_change_cb, NULL, 0,\
            //SR_SUBSCR_ENABLED | SR_SUBSCR_DONE_ONLY| SR_SUBSCR_CTX_REUSE, subscription);
            SR_SUBSCR_ENABLED | SR_SUBSCR_DONE_ONLY , subscription);
    if (iRet != 0) {
        goto error;
    }

    iRet = sr_module_change_subscribe(session, "test", "/test:config/vipPriQWeightList", config_change_cb, NULL, 0,\
            //SR_SUBSCR_ENABLED | SR_SUBSCR_DONE_ONLY| SR_SUBSCR_CTX_REUSE, subscription);
            SR_SUBSCR_ENABLED | SR_SUBSCR_DONE_ONLY , subscription);
    if (iRet != 0) {
        goto error;
    }

    iRet = sr_module_change_subscribe(session, "test", "/test:config/userSchedSrvList", config_change_cb, NULL, 0,\
            //SR_SUBSCR_ENABLED | SR_SUBSCR_DONE_ONLY| SR_SUBSCR_CTX_REUSE, subscription);
            SR_SUBSCR_ENABLED | SR_SUBSCR_DONE_ONLY , subscription);
    if (iRet != 0) {
        goto error;
    }

    /* sysrepo/plugins.h provides an interface for logging */
    return SR_ERR_OK;

error:
    //sr_unsubscribe(subscription);
	return iRet;
```
Please note the `SR_SUBSCR_CTX_REUSE` I have not used because of another issue I submited (mult-subscribe can not trigger callback) ,and I use the `xpath` passed in to distinguish between different lists in the same callback function `config_change_cb()`,in order to execute different process(get different list changes).

This problem seems to have occurred when I was adding or deleting or modifing changes of the list entry.
As I said before, there is no way to recover but to reboot.
I have met it several times, I'm a little skeptical about my usage, and please help me correct it.

**Comment by michalvasko** [comment]
> This problem seems to have occurred when I was adding or deleting or modifing changes of the list entry.

You were doing what exactly? Adding some changes to the existing changes in the callback?

> As I said before, there is no way to recover but to reboot.

Yes, or run `make shm_clean` but you must have no sysrepo applications running.

> I subscribe to multiple callbacks from a container in yang module, but when the data changes comes then triggers the callbacks, from the point of view of printing log, it is multithreaded. What can I do to ensure that my own application executes the callbacks one by one?

Use a distinct priority for each callback for one YANG module.

**Comment by myonlystarWang** [comment]
> You were doing what exactly? Adding some changes to the existing changes in the callback?
I have a webpage UI used for modify/add/delete data,web UI send restful API to _Opendaylight_ when I change the data,and  _Opendaylight_ send netconf messages to netopeer2 in device.

My latest findings looks like the API `sr_xpath_node_name_eq` has something wrong within, I write another API which provide the same functionality(comparison the last string after the last`/`),It has been used for a while without any problems.
I find that the API `sr_xpath_node_name_eq` has not been called in sysrepo code,is it not recommended?

**Comment by michalvasko** [comment]
Those should work, generally, but they were only ported over from the legacy *sysrepo* and are not used anywhere. The code seems quite simple, do you have the non-working use-case?

**Comment by myonlystarWang** [comment]
I'm sorry I didn't.
I can also provide a strange problem when I use the API `sr_xpath_node_name_eq`.I can't necessarily walk into this branch after I call `sr_xpath_node_name_eq`,such as this:
```
                                        if(sr_xpath_node_name_eq(val->xpath, "name")) 
					{
						strcpy(ipsec_name, val->data.string_val);
						has_name = true;
					}
					else if(sr_xpath_node_name_eq(val->xpath, "left")) 
					{
						strcpy(left, val->data.string_val);
						has_left =true;
					}
					else if(sr_xpath_node_name_eq(val->xpath, "right")) 
					{
						strcpy(right, val->data.string_val);
						has_right = true;
					}
```
I really changed the value of `“name”`, but went to `"left"` branches.When this happens, this issue can easily reproduced

**Comment by myonlystarWang** [comment]
> I'm sorry I didn't.
> I can also provide a strange problem when I use the API `sr_xpath_node_name_eq`.I can't necessarily walk into this branch after I call `sr_xpath_node_name_eq`,such as this:
> 
> ```
>                                         if(sr_xpath_node_name_eq(val->xpath, "name")) 
> 					{
> 						strcpy(ipsec_name, val->data.string_val);
> 						has_name = true;
> 					}
> 					else if(sr_xpath_node_name_eq(val->xpath, "left")) 
> 					{
> 						strcpy(left, val->data.string_val);
> 						has_left =true;
> 					}
> 					else if(sr_xpath_node_name_eq(val->xpath, "right")) 
> 					{
> 						strcpy(right, val->data.string_val);
> 						has_right = true;
> 					}
> ```
> 
> I really changed the value of `“name”`, but went to `"left"` branches.When this happens, this issue can easily reproduced

**Comment by myonlystarWang** [comment]
Sorry,I accidentally hit the wrong button

**Comment by michalvasko** [comment]
So what is the value of `val->xpath` when you think it is not working properly? I still do not know any details of the problem.

**Comment by myonlystarWang** [comment]
Ok,Let me make it a little bit clearer.
When I modify the value of `"name"` from webpage,it trigger the callback which I subscribed to sysrepo in device machine.So the following output can be printed:
```
CREATED: /module:container (container)
CREATED: /module:container/name = "aaa"
```
The value of `val->xpath` is  clearly `/module:container/name`,but it can not go into the branch everytime.
```
if(sr_xpath_node_name_eq(val->xpath, "name")) 
{
	strcpy(ipsec_name, val->data.string_val);
	has_name = true;
}
```
It might have walked into this branch for some reason I do not known
```
if(sr_xpath_node_name_eq(val->xpath, "left")) 
{
	strcpy(left, val->data.string_val);
	has_left = true;
}
```
But I suspect something is wrong with the interface `sr_xpath_node_name_eq`

**Comment by michalvasko** [comment]
You are still not helping me much. I need a use-case with which I can reproduce the problem 100% of times, you keep saying that it is sometimes not working. It would make sense if there was some memory problem, for example, so you can try running your app with `valgrind`.

**Comment by myonlystarWang** [comment]
Ok I will try it and keep my eyes open .But in fact it probably will not happen again.

**Comment by myonlystarWang** [comment]
It has never been reproduced again, probably because I used the correct `xpath `comparison interface.
I think I can close the issue now.
